internal/multistorage/storage.go

```diff
@@ -27,46 +28,36 @@ type Config struct {
 	StatusCacheTTL       time.Duration
 }
 
-func NewStorage(
-	config *Config,
-	primary storage.HashableStorage,
-	failovers map[string]storage.HashableStorage,
-) *Storage {
+func NewStorage(config *Config, primary storage.HashableStorage, failovers map[string]storage.HashableStorage) *Storage {
 	specificStorages := NameAndOrderStorages(primary, failovers)
-
-	specificFolders := make(map[string]storage.Folder, len(specificStorages))
-	for _, specSt := range specificStorages {
-		specificFolders[specSt.Name] = specSt.RootFolder()
-	}
-
-	var statusCache cache.StatusCache
-	if config.AliveChecks {
-		var aliveChecker cache.AliveChecker
-		if config.CheckWrite {
-			aliveChecker = cache.NewReadAliveChecker(specificFolders, config.AliveCheckTimeout)
-		} else {
-			aliveChecker = cache.NewRWAliveChecker(specificFolders, config.AliveCheckTimeout, config.AliveCheckWriteBytes)
-		}
-		statusCache = cache.NewStatusCache(storageKeys(specificStorages), config.StatusCacheTTL, aliveChecker)
-	} else {
-		statusCache = cache.NewStatusCacheNoCheck(storageKeys(specificStorages))
-	}
-
-	rootFolder := NewFolder(specificFolders, statusCache)
+	statsCollector := configureStatsCollector(specificStorages, config)
+	rootFolder := NewFolder(specificStorages.RootFolders(), statsCollector)
 
 	return &Storage{
-		statusCache:      statusCache,
+		statsCollector:   statsCollector,
 		specificStorages: specificStorages,
 		rootFolder:       rootFolder,
 	}
 }
 
-func storageKeys(storages []NamedStorage) []cache.Key {
-	res := make([]cache.Key, len(storages))
-	for i, s := range storages {
-		res[i] = s.Key
+func configureStatsCollector(storages NamedStorages, config *Config) stats.Collector {
+	switch {
+	case config.AliveChecks && config.CheckWrite:
+		statusCache := cache.New(storages.Keys(), config.StatusCacheTTL, cache.DefaultRWMem, cache.DefaultRWFile)
+		aliveChecker := stats.NewRWAliveChecker(storages.RootFolders(), config.AliveCheckTimeout, config.AliveCheckWriteBytes)
+		return stats.NewCollector(storages.Names(), statusCache, aliveChecker)
+
+	case config.AliveChecks && !config.CheckWrite:
+		statusCache := cache.New(storages.Keys(), config.StatusCacheTTL, cache.DefaultROMem, cache.DefaultROFile)
+		aliveChecker := stats.NewROAliveChecker(storages.RootFolders(), config.AliveCheckTimeout)
+		return stats.NewCollector(storages.Names(), statusCache, aliveChecker)
+
+	case !config.AliveChecks:
+		return stats.NewNopCollector(storages.Names())
+
+	default:
+		panic("can't init multi-storage stats collector")
```

Comment:
> do we need this panic at all
If we won't panic here or return an error, it could make this problem deferred, which is worse.

> Why we can't return an error here?
That's why I made it panic here - it's very unlikely scenario, so it's like overhead to return and handle the error everywhere higher up the stack.

So, in this case I would perfer to drop this panic at all than to return an error